File: frontend/src/utils/formValidation.ts
import type { FormTemplate } from "../types";

export function validateFormData(
  template: FormTemplate,
  formData: Record<string, unknown>,
  role?: string,
): string[] {
  const errors: string[] = [];
  const scores = (formData.scores as Record<string, number | null | undefined>) ?? {};
  const textFields = (formData.text_fields as Record<string, string | undefined>) ?? {};

  for (const item of template.items) {
    const display = `${item.label} ${item.name}`.trim();
    const score = scores[item.id];
    if (score === undefined || score === null) {
      errors.push(`「${display}」の採点が未入力です`);
    }
  }

  for (const field of template.text_fields ?? []) {
    if (field.readonly) continue;
    if (field.role && role && field.role !== role) continue;
    if (field.role && !role) continue;
    const value = textFields[field.id] ?? "";
    if (!value.trim()) {
      errors.push(`「${field.label}」が未入力です`);
    }
  }

  return errors;
}

export function formatValidationErrors(errors: string[]): string {
  if (errors.length === 0) return "";
  return ["未入力の項目があります:", ...errors.map((e) => `・${e}`)].join("\n");
}

File: backend/app/services/form_validation.py
from typing import Any, Optional


def validate_form_data(
    template: dict[str, Any],
    form_data: dict[str, Any],
    role: Optional[str] = None,
) -> list[str]:
    errors: list[str] = []
    scores = form_data.get("scores") or {}
    text_fields = form_data.get("text_fields") or {}

    for item in template.get("items", []):
        item_id = item["id"]
        label = item.get("label", "")
        name = item.get("name", "")
        display = f"{label} {name}".strip()
        score = scores.get(item_id)
        if score is None or score == "":
            errors.append(f"「{display}」の採点が未入力です")

    for field in template.get("text_fields", []):
        if field.get("readonly"):
            continue
        field_role = field.get("role")
        if field_role and role and field_role != role:
            continue
        if field_role and not role:
            continue
        value = text_fields.get(field["id"], "")
        if not str(value).strip():
            errors.append(f"「{field['label']}」が未入力です")

    return errors
